### PR TITLE
feat: Prove continuity of map K_v -> L_w 

### LINKED ISSUE #204: Continuity of map of local fields K_v -> L_w coming from number fields
Let A be a Dedekind domain with field of fractions K, let L/K be a finite separable extension and let B be the integral closure of A in L.

If w is a finite place of L lying over v of K, then the map K -> L "obviously" extends to a map of completions K_v -> L_w. At the time of writing, Andrew Yang has reduced this claim to the assertion about the relationship between the v-adic and w-adic valuations on K, which should easily follow from `IsDedekindDomain.HeightOneSpectrum.valuation_comap`. The goal of this issue is to check this, i.e. to assume `valuation_comap` and fill in the sorry in `adicCompletion_comap_ringHom` in the file `FLT.DededkindDomain.FiniteAdeleRing.BaseChange`. It should hopefully be a relatively straightforward wrestle with `WithZero (Multiplicative ℤ)`.

### PR DESCRIPTION
**Pull request by javierlcontreras**
Code dictated by Yaël

Closes #204

### CODE REVIEW

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean [hunk 1]
```diff
@@ -60,7 +62,7 @@ def comap (w : HeightOneSpectrum B) : HeightOneSpectrum A where
 lemma map_comap (w : HeightOneSpectrum B) :
     (w.comap A).asIdeal.map (algebraMap A B) =
     w.asIdeal ^ ((Ideal.ramificationIdx (algebraMap A B) (comap A w).asIdeal w.asIdeal)) := by
-  -- This must be standard? Maybe a hole in the library for Dedekind domains
+  -- This mus t be standard? Maybe a hole in the library for Dedekind domains
```
Comment (kbuzzard) [inline FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:65]: remove the space?

Note: Andrew has filled in the sorry.

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean [hunk 2]
```diff
@@ -69,18 +71,23 @@ open scoped algebraMap
 
 
 -- Need to know how the valuation `w` and its pullback are related on elements of `K`.
-def valuation_comap (w : HeightOneSpectrum B) (x : K) :
-    (comap A w).valuation x =
-    w.valuation (algebraMap K L x) ^
-    (Ideal.ramificationIdx (algebraMap A B) (comap A w).asIdeal w.asIdeal) := by
+lemma valuation_comap (w : HeightOneSpectrum B) (x : K) :
+    w.valuation (algebraMap K L x) = (comap A w).valuation x ^
+    (comap A w).asIdeal.ramificationIdx (algebraMap A B) w.asIdeal := by
```
Comment (kbuzzard) [inline FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:76]: Thanks :sweat_smile:

**Review by kbuzzard — COMMENTED** [review]
One of the sorries is false right now, as far as I can see.

**Comment by kbuzzard** [comment]
Can one of you write a message just saying claim in #204 ?

**Comment by YaelDillies** [comment]
If you're against merging with a false sorry, then I suggest we wait a week or two for my refactor(s) to hit mathlib. It's quite inconvenient to work around.

**Comment by kbuzzard** [comment]
I'm definitely against merging with a false sorry! Sorry :-)

**Comment by YaelDillies** [comment]
Actually, this PR has a net count of 1 - 1 = 0 new false sorries ;)

**Comment by kbuzzard** [comment]
Yes but that doesn't stop me objecting to the addition of a new false sorry. In fact the old false sorry (my fault == a sign error) is removed on main already.

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean (hunk 2 above)
Comment (kbuzzard) [reply FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:76]: This is now fixed on `main`.

File: FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import Mathlib.Algebra.Order.Ring.Basic
+
+variable {G₀ : Type*} [GroupWithZero G₀] [LinearOrder G₀]
+variable {a b : G₀}
+
+@[mono, gcongr, bound]
+theorem zpow_le_zpow_left (ha : 0 ≤ a) (hab : a ≤ b) : ∀ n : ℤ, a ^ n ≤ b ^ n :=
+  by sorry
```
Comment (kbuzzard) [inline FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean:8]: This is also false, e.g. 0<=1<=2 but 1^{-1} is not <= 2^{-1}. What's going on in this PR exactly?

File: FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+@[mono, gcongr, bound]
+theorem zpow_lt_zpow_left (ha : 0 ≤ a) (hab : a < b) : ∀ {n : ℤ}, a ^ n < b ^ n :=
+  by sorry
```
Comment (kbuzzard) [inline FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean:12]: As is this.

File: FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+
+theorem lt_of_zpow_lt_zpow_left (n : ℤ) (hb : 0 ≤ b) (h : a^n < b^n) : a < b :=
+  lt_of_not_ge fun hn => not_lt_of_ge (zpow_le_zpow_left hb hn _) h
```
Comment (kbuzzard) [inline FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean:16]: As is this.

File: FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean (hunk 3 above)
Comment (YaelDillies) [reply FLT/Mathlib/Algebra/Order/GroupWithZero/Unbundled.lean:8]: Don't worry. I've already noticed this. It is getting fixed as soon as Javier gives me access to his fork

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean [hunk 6]
```diff
@@ -131,20 +141,17 @@ lemma intValuation_comap (hAB : Function.Injective (algebraMap A B))
...
-    (comap A w).valuation x ^
-    (Ideal.ramificationIdx (algebraMap A B) (comap A w).asIdeal w.asIdeal) =
-    w.valuation (algebraMap K L x) := by
-  obtain ⟨x, y, hy, rfl⟩ := IsFractionRing.div_surjective (A := A) x
-  simp [valuation, ← IsScalarTower.algebraMap_apply A K L, IsScalarTower.algebraMap_apply A B L,
-    ← intValuation_comap A B (algebraMap_injective_of_field_isFractionRing A B K L), div_pow]
+    w.valuation (algebraMap K L x) = (comap A w).valuation x ^
+    (comap A w).asIdeal.ramificationIdx (algebraMap A B) w.asIdeal := by
+  -- This should follow from map_comap?
+  sorry
```
Comment (kbuzzard) [inline FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:148]: Why are you deleting a completed proof and replacing it with a `sorry`?

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean [hunk 7]
```diff
@@ -61,6 +63,14 @@ def comap (w : HeightOneSpectrum B) : HeightOneSpectrum A where
   isPrime := Ideal.comap_isPrime (algebraMap A B) w.asIdeal
   ne_bot := mt Ideal.eq_bot_of_comap_eq_bot w.ne_bot
 
+-- lemma: pushforward of pullback is P^(ram index)
+lemma map_comap (w : HeightOneSpectrum B) :
+    (w.comap A).asIdeal.map (algebraMap A B) =
+    w.asIdeal ^ ((Ideal.ramificationIdx (algebraMap A B) (comap A w).asIdeal w.asIdeal)) := by
+  -- This mus t be standard? Maybe a hole in the library for Dedekind domains
+  -- or maybe I just missed it?
+  sorry
```
Comment (kbuzzard) [inline FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:72]: Why are you adding this sorried lemma?

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean [hunk 8]
```diff
@@ -131,20 +141,17 @@ lemma intValuation_comap (hAB : Function.Injective (algebraMap A B))
...
+    (comap A w).asIdeal.ramificationIdx (algebraMap A B) w.asIdeal := by
+  -- This should follow from map_comap?
+  sorry
 
 -- Say w is a finite place of L lying above v a finite places
 -- of K. Then there's a ring hom K_v -> L_w.
 variable {B L} in
-noncomputable def adicCompletion_comap_ringHom (w : HeightOneSpectrum B) :
-    (adicCompletion K (comap A w)) →+* (adicCompletion L w) :=
+noncomputable def adicCompletion_comap (w : HeightOneSpectrum B) :
```
Comment (kbuzzard) [inline FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:153]: Why are you changing the name? _comap` should be the algebra hom, not this.

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean (hunk 6 above)
Comment (YaelDillies) [reply FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:148]: Whoops, looks like a bad merge

**Comment by YaelDillies** [comment]
Somehow when I merged master it overwrote the new version of the file without telling me. I'll revert what I spot

File: FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean [hunk 9]
```diff
@@ -1,7 +1,7 @@
-import Mathlib -- **TODO** fix when finished or if `exact?` is too slow
---import Mathlib.RingTheory.DedekindDomain.FiniteAdeleRing
---import Mathlib.NumberTheory.NumberField.Basic
---import Mathlib.NumberTheory.RamificationInertia
+import Mathlib
+import FLT.Mathlib.Algebra.Order.Monoid.Unbundled.TypeTags
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply FLT/DedekindDomain/FiniteAdeleRing/BaseChange.lean:5]: Again, bad merge, but you know about `#min_imports`, right?

File: FLT/Mathlib/Algebra/Order/Hom/Monoid.lean
Comment (YaelDillies) [inline FLT/Mathlib/Algebra/Order/Hom/Monoid.lean:1]: This is now in mathlib, so can disappear if you bump

Comment (kbuzzard) [reply FLT/Mathlib/Algebra/Order/Hom/Monoid.lean:1]: Bumped!

Comment (YaelDillies) [reply FLT/Mathlib/Algebra/Order/Hom/Monoid.lean:1]: Removed

**Comment by kbuzzard** [comment]
Thanks! Sorry for the delay, I've been away all week.

**Pull request merged**